File: examples/typescript/core/api.case.ts
/* ========================================================================== *
 * APP v0.0.3
 * core/api.case.ts
 * ----------------------------------------------------------------------------
 * Contrato base da surface de API no APP.
 *
 * Responsabilidade:
 * - expor uma capacidade via interface de backend (HTTP, RPC, CLI, etc)
 * - orquestrar validação, autorização e execução
 * - retornar uma resposta estruturada
 *
 * Regra fundamental:
 * - lógica de domínio pertence ao domain.case.ts
 * - persistência ou integração deve ser encapsulada em métodos privados
 *
 * Contexto:
 * - ApiContext estende AppBaseContext com infraestrutura de backend
 * - cada projeto define os tipos concretos de http, db, auth, etc.
 * ========================================================================== */

import { Dict } from "./domain.case";
import { AppBaseContext } from "./shared/app_base_context";
import { AppHttpClient, AppStorageClient, AppCache } from "./shared/app_infra_contracts";
import { AppCaseError, AppResult } from "./shared/app_structural_contracts";

/* ==========================================================================
 * ApiContext
 * --------------------------------------------------------------------------
 * Contexto específico da surface de API.
 *
 * Estende AppBaseContext com infraestrutura de backend:
 * - httpClient: outbound HTTP client (fetch, axios, etc.)
 * - db: acesso a banco de dados (unknown — sem contrato estável)
 * - auth: autenticação e autorização (unknown — semântica de domínio)
 * - storage: persistent storage client
 * - cache: cache com TTL
 *
 * Campos com contrato mínimo usam interfaces de app_infra_contracts.ts.
 * Campos sem semântica estável permanecem unknown.
 * ========================================================================== */

export interface ApiContext extends AppBaseContext {
  /**
   * Outbound HTTP client.
   *
   * Exemplos: fetch wrapper, Axios instance, got, ky, undici.
   * Nota: este é um client contract, não um server/framework contract.
   */
  httpClient?: AppHttpClient;

  /**
   * Acesso a banco de dados.
   *
   * Mantido como unknown — paradigmas incompatíveis (ORM, query builder,
   * document store) impedem contrato mínimo convergente.
   *
   * Exemplos: Prisma client, Drizzle, Knex, connection pool.
   */
  db?: unknown;

  /**
   * Informações de autenticação e autorização.
   *
   * Mantido como unknown — carrega semântica de domínio que varia
   * entre modelos (RBAC, ABAC, claims, scopes, sessions).
   *
   * Exemplos: JWT decoded, session object, API key metadata.
   */
  auth?: unknown;

  /**
   * Persistent storage client.
   *
   * Exemplos: S3 client, GCS client, local filesystem adapter.
   */
  storage?: AppStorageClient;

  /**
   * Cache com TTL opcional.
   *
   * Exemplos: Redis client, in-memory cache, Memcached.
   */
  cache?: AppCache;

  /**
   * Registro de Cases carregados pelo runtime.
   *
   * Permite composição cross-case via registry boundary.
   * Usado por `_composition` para resolver capabilities de outros Cases.
   *
   * Exemplo: ctx.cases?.users?.user_validate?.api?.handler(input)
   */
  cases?: Dict;

  /**
   * Espaço de extensão livre para o host do projeto.
   */
  extra?: Dict;
}

/* ==========================================================================
 * ApiResponse
 * --------------------------------------------------------------------------
 * Extends AppResult with API-specific metadata.
 *
 * ApiResponse inherits the canonical result shape (success, data, error)
 * and adds optional HTTP-specific fields. Surfaces that don't need
 * API-specific metadata can use AppResult directly.
 * ========================================================================== */

/**
 * Estrutura de resposta da surface de API.
 *
 * Estende AppResult com metadados opcionais específicos de API.
 */
export interface ApiResponse<T = unknown> extends AppResult<T> {
  /**
   * HTTP status code hint.
   *
   * The runtime/adapter may use this to set the HTTP response status.
   * APP does not mandate HTTP — this is a convenience for HTTP-based hosts.
   */
  statusCode?: number;
}

/* ==========================================================================
 * BaseApiCase
 * ========================================================================== */

/**
 * Classe base para surfaces de API.
 */
export abstract class BaseApiCase<TInput = unknown, TOutput = unknown> {
  protected readonly ctx: ApiContext;

  constructor(ctx: ApiContext) {
    this.ctx = ctx;
  }

  /* =======================================================================
   * Métodos obrigatórios
   * ===================================================================== */

  /**
   * Handler principal da capacidade.
   *
   * handler é o entrypoint público da capability. Recebe input de negócio
   * e retorna resultado de negócio. NÃO é um endpoint HTTP.
   *
   * Bindings de transporte (HTTP routes, gRPC definitions, CLI commands)
   * vivem em router() ou no adapter/host.
   *
   * Deve:
   * - validar input
   * - verificar autorização
   * - executar lógica principal
   * - retornar resposta estruturada
   */
  public abstract handler(input: TInput): Promise<ApiResponse<TOutput>>;

  /**
   * Router opcional — bindings de transporte.
   *
   * É onde o Case declara sua superfície de transporte (HTTP, gRPC, CLI).
   * O router delega para handler(), nunca contém lógica de negócio.
   * O host/adapter monta as rotas coletando os router() de cada Case.
   *
   * Retorno é framework-specific (unknown).
   */
  public router?(): unknown;

  /**
   * Teste interno da capacidade.
   *
   * Obrigatório no APP — toda surface que implementa um contrato base
   * deve fornecer um método test().
   *
   * Assinatura canônica: test(): Promise<void>
   * O teste invoca handler() internamente e faz assertions.
   * Não recebe input nem retorna resultado — é validação interna.
   */
  public abstract test(): Promise<void>;

  /* =======================================================================
   * Hooks protegidos (opcionais)
   * ===================================================================== */

  /**
   * Validação de input antes da execução.
   */
  protected async _validate?(input: TInput): Promise<void>;

  /**
   * Verificação de autorização.
   */
  protected async _authorize?(input: TInput): Promise<void>;

  /**
   * Acesso a persistência e integrations locais do Case.
   *
   * Slot canônico para queries, mutations, cache reads e chamadas
   * a serviços externos de infraestrutura.
   *
   * Regra: _repository não realiza composição cross-case.
   */
  protected _repository?(): unknown;

  /**
   * Execução da lógica principal (Case atômico).
   *
   * Slot canônico para lógica de negócio que não envolve
   * orquestração cross-case. Mutuamente exclusivo com _composition
   * como centro de execução principal.
   *
   * Cases atômicos implementam _service.
   * Cases compostos implementam _composition.
   * O pipeline exige que ao menos um dos dois esteja definido.
   */
  protected async _service?(input: TInput): Promise<TOutput>;

  /**
   * Orquestração cross-case via registry (Case composto).
   *
   * Slot canônico para Cases que precisam invocar outros Cases.
   * Resolve capabilities via ctx.cases, nunca por import direto.
   *
   * Mutuamente exclusivo com _service como centro de execução principal.
   * Quando presente, o pipeline deve usar _composition em vez de _service.
   */
  protected async _composition?(input: TInput): Promise<TOutput>;

  /**
   * Método utilitário padrão para execução.
   *
   * Orquestra o pipeline: validate → authorize → (composition | service).
   *
   * Se _composition estiver definido, ele é o centro de execução (Case composto).
   * Caso contrário, _service é usado (Case atômico).
   */
  protected async execute(input: TInput): Promise<ApiResponse<TOutput>> {
    try {
      if (this._validate) await this._validate(input);
      if (this._authorize) await this._authorize(input);

      if (!this._composition && !this._service) {
        throw new AppCaseError(
          "INTERNAL",
          "BaseApiCase: at least one of _service or _composition must be implemented"
        );
      }

      const result = this._composition
        ? await this._composition(input)
        : await this._service!(input);

      return {
        success: true,
        data: result,
      };
    } catch (err) {
      if (err instanceof AppCaseError) {
        return {
          success: false,
          error: err.toAppError(),
        };
      }
      // Unexpected errors re-throw — the host/adapter decides how to handle them.
      throw err;
    }
  }
}

File: examples/python/core/api_case.py
# ==========================================================================
# APP v1.1.4
# core/api_case.py
# --------------------------------------------------------------------------
# Base contract for the API surface in APP.
#
# Responsibility:
# - expose a capability via a backend interface (HTTP, RPC, CLI, etc.)
# - orchestrate validation, authorization, and execution
# - return a structured response
#
# Fundamental rule:
# - domain logic belongs in domain_case.py
# - persistence or integration must be encapsulated in private methods
#
# Context:
# - ApiContext extends AppBaseContext with backend infrastructure
# - each project defines the concrete types for http, db, auth, etc.
# ==========================================================================

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

from core.shared.app_base_context import AppBaseContext
from core.shared.app_infra_contracts import (
    AppCache,
    AppHttpClient,
    AppStorageClient,
)
from core.shared.app_structural_contracts import (
    AppCaseError,
    AppError,
    AppResult,
)

TInput = TypeVar("TInput")
TOutput = TypeVar("TOutput")


# ==========================================================================
# ApiContext
# --------------------------------------------------------------------------
# Context specific to the API surface.
#
# Extends AppBaseContext with backend infrastructure:
# - http_client: outbound HTTP client
# - db: database access (Any — no stable contract)
# - auth: authentication and authorization (Any — domain semantics)
# - storage: persistent storage client
# - cache: cache with TTL
# - cases: Cases loaded by the runtime (for cross-case composition)
# - packages: library packages registered by the host
# ==========================================================================

class ApiContext(AppBaseContext, total=False):
    """Context for the API surface."""

    http_client: AppHttpClient
    db: Any
    auth: Any
    storage: AppStorageClient
    cache: AppCache
    cases: dict[str, Any]
    packages: dict[str, Any]
    extra: dict[str, Any]


# ==========================================================================
# ApiResponse
# --------------------------------------------------------------------------
# Extends AppResult with API-specific metadata.
# ==========================================================================

@dataclass
class ApiResponse(AppResult[TOutput], Generic[TOutput]):
    """API response with optional HTTP status code hint."""

    # HTTP status code hint. The runtime/adapter may use this.
    status_code: int | None = None


# ==========================================================================
# BaseApiCase
# ==========================================================================

class BaseApiCase(ABC, Generic[TInput, TOutput]):
    """Base class for API surfaces in APP."""

    def __init__(self, ctx: ApiContext) -> None:
        self.ctx = ctx

    # ==================================================================
    # Required methods
    # ==================================================================

    @abstractmethod
    def handler(self, input: TInput) -> ApiResponse[TOutput]:
        """
        Main capability handler.

        Receives business input and returns a business result.
        NOT an HTTP endpoint — transport bindings live in router().

        Should: validate, authorize, execute, return structured response.
        """
        ...

    def router(self) -> Any:
        """
        Optional transport bindings.

        Where the Case declares its transport surface (HTTP, gRPC, CLI).
        The router delegates to handler(), never contains business logic.
        The host/adapter collects router() from each Case to mount routes.

        Return type is framework-specific.
        """
        return None

    def test(self) -> None:
        """
        Self-contained API surface test.

        Canonical signature: test() -> None
        Invokes handler() internally and makes assertions.
        """

    # ==================================================================
    # Protected hooks (optional)
    # ==================================================================

    def _validate(self, input: TInput) -> None:
        """Validate input before execution."""

    def _authorize(self, input: TInput) -> None:
        """Check authorization."""

    def _repository(self) -> Any:
        """
        Persistence and local integrations.

        Canonical slot for queries, mutations, cache reads, and calls
        to external infrastructure services.

        Rule: _repository does not perform cross-case composition.
        """
        return None

    def _service(self, input: TInput) -> TOutput | None:
        """
        Main execution logic (atomic Case).

        Canonical slot for business logic that does not involve
        cross-case orchestration. Mutually exclusive with _composition
        as the main execution center.
        """
        return None

    def _composition(self, input: TInput) -> TOutput | None:
        """
        Cross-case orchestration via registry (composed Case).

        Canonical slot for Cases that need to invoke other Cases.
        Resolves capabilities via ctx.cases, never by direct import.

        Mutually exclusive with _service as the main execution center.
        """
        return None

    # ==================================================================
    # Execution pipeline utility
    # ==================================================================

    def execute(self, input: TInput) -> ApiResponse[TOutput]:
        """
        Standard execution pipeline.

        Orchestrates: validate -> authorize -> (composition | service).

        If _composition returns non-None, it is the execution center (composed).
        Otherwise, _service is used (atomic).
        """
        try:
            self._validate(input)
            self._authorize(input)

            composition_result = self._composition(input)
            if composition_result is not None:
                return ApiResponse(success=True, data=composition_result)

            service_result = self._service(input)
            if service_result is not None:
                return ApiResponse(success=True, data=service_result)

            raise AppCaseError(
                "INTERNAL",
                "BaseApiCase: at least one of _service or _composition must be implemented",
            )
        except AppCaseError as err:
            return ApiResponse(success=False, error=err.to_app_error())
